# Project: Wrangle and Analyze Data

# Gathering

We can gather from 3 different sources, and 3 different types:
- The WeRateDogs Twitter Archive (csv)
- The Tweet Image Predictions (tsv)
- Tweet's retweet count and favorite (Twitter API) (txt)


In [2]:
# Importing our needed libraries
import pandas as pd
import numpy as np
import requests

## 1. We Rate Dogs Archive
This is given to us by Udacity, it is named: `twitter_archive_enhanced.csv`
I have downloaded from **'project description'**

In [3]:
# loading in the file
df_twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')


## 2. Tweet Image Predictions
This is also given to us by Udacity, however we must use the `Requests` library to download it from a given URL.

In [4]:
# downloading files using requests
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)

open('image-predictions.tsv','wb').write(r.content)


335079

In [5]:
# load the tsv into a dataframe (making sure to seperate by tab)
df_predicted_images = pd.read_csv('image-predictions.tsv', sep = '\t')

## 3. Querying the Twitter API
* Grabbing every tweet's **Retweet** and **Favorite** count using `Tweepy`

    Steps:
    
    * Use tweet IDs from the archive
    * Query the twitter API for each tweet's JSON
    * Store the data in a file called `tweet_json.txt`
    * Each JSON should be written to its own line
    * read the .txt file line by line into pandas DF storing:
        * Tweet ID
        * Retweet Count
        * Favorite Count
        
*note: the phone verificaiton has some trouble, so I went ahead and used the file provided by Udacity, I will double check the verification in the future.*

In [6]:
# # BIG NOTE: Oddly, this is also resulting in Fails instead of the real value, so I have uploaded the tweet-json.txt file 
# # which was included next to the code.

# import tweepy
# from tweepy import OAuthHandler
# import json
# from timeit import default_timer as timer

# # Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# # These are hidden to comply with Twitter's API terms and conditions
# consumer_key = 'HIDDEN'
# consumer_secret = 'HIDDEN'
# access_token = 'HIDDEN'
# access_secret = 'HIDDEN'

# auth = OAuthHandler(consumer_key, consumer_secret)
# auth.set_access_token(access_token, access_secret)

# api = tweepy.API(auth, wait_on_rate_limit=True)

# # NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# # df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# # change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# # NOTE TO REVIEWER: this student had mobile verification issues so the following
# # Twitter API code was sent to this student from a Udacity instructor
# # Tweet IDs for which to gather additional data via Twitter's API
# tweet_ids = df_twitter_archive.tweet_id.values
# len(tweet_ids)

# # Query Twitter's API for JSON data for each tweet ID in the Twitter archive
# count = 0
# fails_dict = {}
# start = timer()
# # Save each tweet's returned JSON as a new line in a .txt file
# with open('tweet_json.txt', 'w') as outfile:
#     # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
#     for tweet_id in tweet_ids:
#         count += 1
#         print(str(count) + ": " + str(tweet_id))
#         try:
#             tweet = api.get_status(tweet_id, tweet_mode='extended')
#             print("Success")
#             json.dump(tweet._json, outfile)
#             outfile.write('\n')
#         except tweepy.TweepError as e:
#             print("Fail")
#             fails_dict[tweet_id] = e
#             pass
# end = timer()
# print(end - start)
# print(fails_dict)

In [7]:
# link for this useful code: https://stackoverflow.com/questions/30088006/loading-a-file-with-more-than-one-line-of-json-into-pandas

df_tweets = pd.read_json('tweet-json.txt',lines=True)

# Assessing

We will be assessing every file we got above in the same order:
1. We Rate Dogs Archive
2. Tweet Image Predictions
3. Querying the Twitter API

Keys of assessment:
* Completeness
* Validity
* Accuracy
* Consistency 

at the end we will group the issues under Quality and Tidiness

## Asessing We Rate Dogs Archive

### Visual Assesment
using pandas .head() and .sample()

In [8]:
# It's best to simply look at the head
df_twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [9]:
# It's also beneficial to look at a sample
df_twitter_archive.sample(25)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2002  672481316919734272                    NaN                  NaN   
2107  670465786746662913                    NaN                  NaN   
1327  705975130514706432                    NaN                  NaN   
1205  715928423106027520                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
1655  683391852557561860                    NaN                  NaN   
766   777684233540206592                    NaN                  NaN   
1586  686760001961103360                    NaN                  NaN   
1291  708119489313951744                    NaN                  NaN   
1839  675891555769696257                    NaN                  NaN   
1559  688547210804498433                    NaN                  NaN   
731   781655249211752448                    NaN                  NaN   
661   791026214425268224                    NaN                  NaN   
161   860563773140209665                    NaN                  NaN   
2016  672169685991993344                    NaN                  NaN   
546   805823200554876929                    NaN                  NaN   
168   859607811541651456                    NaN                  NaN   
1901  674646392044941312                    NaN                  NaN   
1597  686050296934563840                    NaN                  NaN   
150   863062471531167744                    NaN                  NaN   
1563  688179443353796608                    NaN                  NaN   
1885  674793399141146624           6.717299e+17         4.196984e+09   
1917  674291837063053312                    NaN                  NaN   
1422  698178924120031232                    NaN                  NaN   
686   788070120937619456                    NaN                  NaN   

                      timestamp  \
2002  2015-12-03 18:23:34 +0000   
2107  2015-11-28 04:54:34 +0000   
1327  2016-03-05 04:36:02 +0000   
1205  2016-04-01 15:46:52 +0000   
211   2017-04-12 00:23:33 +0000   
1655  2016-01-02 20:58:09 +0000   
766   2016-09-19 01:42:24 +0000   
1586  2016-01-12 04:01:58 +0000   
1291  2016-03-11 02:36:57 +0000   
1839  2015-12-13 04:14:39 +0000   
1559  2016-01-17 02:23:42 +0000   
731   2016-09-30 00:41:48 +0000   
661   2016-10-25 21:18:40 +0000   
161   2017-05-05 18:36:06 +0000   
2016  2015-12-02 21:45:16 +0000   
546   2016-12-05 17:16:37 +0000   
168   2017-05-03 03:17:27 +0000   
1901  2015-12-09 17:46:48 +0000   
1597  2016-01-10 05:01:51 +0000   
150   2017-05-12 16:05:02 +0000   
1563  2016-01-16 02:02:19 +0000   
1885  2015-12-10 03:30:58 +0000   
1917  2015-12-08 18:17:56 +0000   
1422  2016-02-12 16:16:41 +0000   
686   2016-10-17 17:32:13 +0000   

                                                 source  \
2002  <a href="http://twitter.com/download/iphone" r...   
2107  <a href="http://twitter.com/download/iphone" r...   
1327  <a href="http://twitter.com/download/iphone" r...   
1205  <a href="http://twitter.com/download/iphone" r...   
211   <a href="http://twitter.com/download/iphone" r...   
1655  <a href="http://twitter.com/download/iphone" r...   
766   <a href="http://twitter.com/download/iphone" r...   
1586  <a href="http://vine.co" rel="nofollow">Vine -...   
1291  <a href="http://twitter.com/download/iphone" r...   
1839  <a href="http://twitter.com/download/iphone" r...   
1559  <a href="http://twitter.com/download/iphone" r...   
731   <a href="http://vine.co" rel="nofollow">Vine -...   
661   <a href="http://twitter.com/download/iphone" r...   
161   <a href="http://twitter.com/download/iphone" r...   
2016  <a href="http://twitter.com/download/iphone" r...   
546   <a href="http://twitter.com/download/iphone" r...   
168   <a href="http://twitter.com/download/iphone" r...   
1901  <a href="http://twitter.com/download/iphone" r...   
1597  <a href="http://twitter.com/download/iphone" r...   
150   <a href="http://twitter.com/download/iphone" r

### Programmatic assessment

In [10]:
# Checking .info
df_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [11]:
# checking for duplicates on the tweet id
df_twitter_archive.tweet_id.duplicated().sum()

0

In [12]:
# checking for denominators that arent 10
np.sort(df_twitter_archive.rating_denominator.unique())

array([  0,   2,   7,  10,  11,  15,  16,  20,  40,  50,  70,  80,  90,
       110, 120, 130, 150, 170])

In [13]:
# checking the number of occurances
df_twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

**From what we can see, is that the number of occurances for these numbers are so low, and from the information given in the project details we can say that these might be inforrectly gathered so we have to manually check them.**

In [17]:
# we use this line to print the whole text:
pd.set_option("display.max_colwidth", -1)

# check the rows where the denomintor is not 10:
df_twitter_archive.query('rating_denominator != "10"')[['rating_numerator','rating_denominator']]'rating_numerator','rating_denominator'

rating_numerator  rating_denominator
313   960               0                 
342   11                15                
433   84                70                
516   24                7                 
784   9                 11                
902   165               150               
1068  9                 11                
1120  204               170               
1165  4                 20                
1202  50                50                
1228  99                90                
1254  80                80                
1274  45                50                
1351  60                50                
1433  44                40                
1598  4                 20                
1634  143               130               
1635  121               110               
1662  7                 11                
1663  20                16                
1779  144               120               
1843  88                80                
2335  1                 2

In [18]:
# getting the text for the rows printed above: 
df_twitter_archive.query('rating_denominator != "10"').text

313     @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho                                                                       
342     @docmisterio account started on 11/15/15                                                                                                                                 
433     The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                      
516     Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx
784     RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…                             
902     Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                        

In [ ]:
df_twitter_archive.query('rating_denominator != "10"').expanded_urls

## Quality Issues

### We Rate Dogs Archive

* Some ratings have been gathered incorrectly (nominator and denominator) the indices and their problems:

| index| Incorrect | Correct |
| --- | --- | --- |
| 313 | 960 / 0 | 13 / 10 |  
| 342 | 11/15 | NO RATING |
| 516 | 24/7  | NO RATING |
| 

* 342 is not a rating tweet.

* Missing Expanded URLS 

### Tweet Image Predictions

*
*

### Querying the Twitter API

*
*



## Tidiness Issues

### We Rate Dogs Archive

* doggo, floofer, pupper, and puppo (one variable 4 columns)

*

### Tweet Image Predictions

*
*

### Querying the Twitter API

*
*

# Cleaning

# Storing Wrangled Data